presto-main/src/main/java/com/facebook/presto/operator/aggregation/differentialentropy/WeightedReservoirSampleStateStrategy.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.differentialentropy;
+
+import com.facebook.presto.operator.aggregation.reservoirsample.WeightedDoubleReservoirSample;
+import com.facebook.presto.spi.PrestoException;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.SliceOutput;
+
+import static com.facebook.presto.operator.aggregation.differentialentropy.EntropyCalculations.calculateFromSamples;
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.google.common.base.Verify.verify;
+import static java.lang.String.format;
+
+public class WeightedReservoirSampleStateStrategy
+        implements DifferentialEntropyStateStrategy
+{
+    private final WeightedDoubleReservoirSample reservoir;
+
+    public WeightedReservoirSampleStateStrategy(long maxSamples)
+    {
+        if (maxSamples <= 0) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("maxSamples must be positive: %s", maxSamples));
+        }
+        if (maxSamples >= WeightedDoubleReservoirSample.MAX_SAMPLES_LIMIT) {
+            throw new PrestoException(
+                    INVALID_FUNCTION_ARGUMENT,
+                    format("maxSamples must be capped: %s, %s", WeightedDoubleReservoirSample.MAX_SAMPLES_LIMIT));
+        }
+
+        reservoir = new WeightedDoubleReservoirSample((int) maxSamples);
+    }
+
+    private WeightedReservoirSampleStateStrategy(WeightedReservoirSampleStateStrategy other)
+    {
+        reservoir = other.reservoir.clone();
+    }
+
+    private WeightedReservoirSampleStateStrategy(WeightedDoubleReservoirSample reservoir)
+    {
+        this.reservoir = reservoir;
+    }
+
+    @Override
+    public void validateParameters(
```

Comment:
remove this method

Reply:
I really don't understand why. The principles of OO, AFAIU, are that you want to minimize explicit RTTI, by having virtual methods specialized by type. Exceptions are only when there is no object, like in a polymorphic factory, or more than a single method, like in multiple dispatch.